pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,197 @@
+import random
+
+import torch
+from torch.autograd import Variable
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.util import torch_zeros_like, torch_ones_like
+from pyro.infer import TracePosterior
+from pyro.poutine import Trace
+from tests.common import assert_equal
+
+
+def verlet_integrator(z, r, grad_potential, step_size, num_steps):
+    """
+    Velocity Verlet integrator.
+
+    :param z: trace object containing current values for the sample sites
```

Comment:
This `verlet_integrator()` would be more valuable if it were less dependent on `Trace` objects, say if `z` and `r` were both simple python dicts of `torch.autograd.Variable`s. If we could eliminate all Pyro dependencies, then `verlet_integrator()` could be reused outside of the context of tracing. Moreover it would be possible to unit test without a `Trace` or `HMC` object.

I believe this could be accomplished by adding a little complexity to `HMC.log_pdf()`, say a `self.prototype_trace` onto which `z` could be spliced prior to `replay`. You've already done much of the work by abstracting out `grad_potential`.

Reply:
Cool, will give this a try. I initially had a `dict` for `z`, but had put in some complexity in `ReplayPoutine` itself. I think this should be a relatively light weight wrapper, and we can do this within this class itself.